src/IO/H5/VolumeData.cpp

```diff
@@ -154,6 +162,308 @@ void append_element_extents_and_connectivity(
+    quadrature_array[i] = Spectral::to_quadrature(element_quadratures[i]);
+    extents_array[i] = element_extents[i];
+  }
+
+  return Mesh<SpatialDim>{extents_array, basis_array, quadrature_array};
+}
+
+template <size_t SpatialDim>
+std::vector<std::array<double, SpatialDim>> generate_block_logical_coordinates(
+    const std::vector<std::array<double, SpatialDim>>&
+        element_logical_coordinates,
+    const std::string& grid_name,
+    const std::array<int, SpatialDim>& h_refinement_array) {
+  size_t grid_points_x_start_position = 0;
+  std::vector<std::array<double, SpatialDim>> block_logical_coordinates;
+  block_logical_coordinates.reserve(element_logical_coordinates.size());
+  std::vector<double> number_of_elements_each_direction;
+  number_of_elements_each_direction.reserve(SpatialDim);
+  std::vector<double> shift_each_direction;
+  shift_each_direction.reserve(SpatialDim);
+
+  for (size_t i = 0; i < SpatialDim; ++i) {
+    double number_of_elements = pow(2, h_refinement_array[i]);
+    number_of_elements_each_direction.push_back(number_of_elements);
+    size_t grid_points_start_position =
+        grid_name.find('I', grid_points_x_start_position + 1);
+    size_t grid_points_end_position =
+        grid_name.find(',', grid_points_start_position);
+    if (i == SpatialDim) {
+      grid_points_end_position =
+          grid_name.find(')', grid_points_start_position);
+    }
+    int element_index = std::stoi(grid_name.substr(
+        grid_points_start_position + 1,
+        grid_points_end_position - grid_points_start_position - 1));
+    double shift = (-1 + (2 * element_index + 1) / number_of_elements);
+    shift_each_direction.push_back(shift);
+    grid_points_x_start_position = grid_points_start_position;
+  }
+
+  for (size_t i = 0; i < element_logical_coordinates.size(); ++i) {
+    std::array<double, SpatialDim> grid_point_coordinate = {};
+    for (size_t j = 0; j < grid_point_coordinate.size(); ++j) {
+      grid_point_coordinate[j] = 1 / number_of_elements_each_direction[j] *
+                                     element_logical_coordinates[i][j] +
+                                 shift_each_direction[j];
+    }
+    block_logical_coordinates.push_back(grid_point_coordinate);
+  }
+
+  return block_logical_coordinates;
+}
+
+// Sorting routine for an incoming list of values
+std::vector<double> sort_and_order(std::vector<double>& unsorted_coordinate) {
+  std::vector<double> sorted_coordinate;
+  // Come up with way to figure out unsorted_coordinate length. Maybe number of
+  // grid points???
+  sort(unsorted_coordinate.begin(), unsorted_coordinate.end());
+  sorted_coordinate.push_back(unsorted_coordinate[0]);
+  for (size_t i = 1; i < unsorted_coordinate.size(); ++i) {
+    if (unsorted_coordinate[i] == sorted_coordinate.end()[-1]) {
+      continue;
+    } else {
+      sorted_coordinate.push_back(unsorted_coordinate[i]);
+    }
+  }
+  return sorted_coordinate;
+}
+
+// Builds the connectivity by cube
```

Comment:
Maybe give some comments sketching out the algorithm? Connectivity is one of the more complicated parts of the code (maybe somewhat surprisingly)

Reply:
ping?